### PR TITLE
Created disabler object type

### PR DESCRIPTION
**Pull request by jpakkane**
Extracted from Dolphin options.

### CODE REVIEW

**Comment by nirbheek** [comment]
I haven't done a full review yet (quickly skimmed the patch), why doesn't `dependency(..., required : false)` return a disabler when the dependency is not found?

**Comment by jpakkane** [comment]
That would be a backwards-incompatible change in behaviour.

**Comment by nirbheek** [comment]
Oh, of course, the dolphin bits are needed to get that context. Sorry.

File: docs/markdown/Disabler.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+```meson
+f = d.found() # returns false
+```
+
+Any other statement that uses a disabler object will immediately
+return a disabler. For example assuming that `d` contains a disabler
+object then
+
+```meson
+d2 = some_func(d) # value of d will be disabler
```
Comment (bluetech) [inline docs/markdown/Disabler.md]: `d` -> `d2` in the comment? (Also below).

File: docs/markdown/Disabler.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+d3 = true or dr   # value of d will be disabler
```
Comment (bluetech) [inline docs/markdown/Disabler.md]: `dr` -> `d`?

File: docs/markdown/Disabler.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  dependencies : dep)
+
+# And ín a third directory
+
+exe = executable('mytest', 'mytest.c',
+  link_with : lib)
+test('mytest', exe)
+```
+
+This works fine but gets a bit unflexible when you want to make this
```
Comment (jon-turney) [inline docs/markdown/Disabler.md]: 'inflexible' is far more commonly used in idiomatic english

File: mesonbuild/coredata.py [hunk 4]
```diff
@@ -125,6 +125,7 @@ def validate_value(self, value):
             raise MesonException('Value %s not one of accepted values.' % value)
         return value
 
+
```
Comment (jon-turney) [inline mesonbuild/coredata.py]: whitespace change

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -2142,7 +2143,6 @@ def _find_cached_dep(self, name, kwargs):
                     break
         return identifier, cached_dep
 
-    @permittedKwargs(permitted_kwargs['dependency'])
```
Comment (jon-turney) [inline mesonbuild/interpreter.py]: Seems unrelated?

File: mesonbuild/interpreterbase.py [hunk 6]
```diff
@@ -100,6 +100,29 @@ class MutableInterpreterObject(InterpreterObject):
...
+    for i in args:
+        if isinstance(i, Disabler):
+            return True
+    for i in kwargs.values():
+        if isinstance(i, Disabler):
+            return True
+        if isinstance(i, list):
+            for j in i:
+                if isinstance(j, Disabler):
+                    return True
```
Comment (jon-turney) [inline mesonbuild/interpreterbase.py]: Is this going to work correctly with nested lists?  Or does it need recurse or flatten the list?

File: mesonbuild/interpreterbase.py [hunk 7]
```diff
@@ -404,18 +458,26 @@ def method_call(self, node):
         if isinstance(obj, int):
             return self.int_method_call(obj, method_name, args)
         if isinstance(obj, list):
-            return self.array_method_call(obj, method_name, self.reduce_arguments(args)[0])
+            return self.array_method_call(obj, method_name, args)
```
Comment (jon-turney) [inline mesonbuild/interpreterbase.py]: Seems unrelated?

Comment (jpakkane) [reply mesonbuild/interpreterbase.py]: I made these all behave the same. There's no reason to keep the inconsistency and doing a full MR for this change seems excessive.
